labrador/src/poseidon.rs

```diff
@@ -0,0 +1,191 @@
+//
+// Sponge Operation:
+// - 'rate' for input/output, 'capacity' for security.
+// - 'permute' mixes the state using rounds.
+//
+// Functions:
+// - new(): Initializes the sponge.
+// - apply_ark(), apply_s_box(), apply_mds(): Round operations.
+// - permute(): Applies Poseidon permutation.
+// - absorb(): Absorbs input.
+// - squeeze(): Squeezes output.
+//
+
+use crate::zq::Zq;
+
+#[derive(Clone, Debug)]
+/// Struct for the Poseidon sponge
+pub struct PoseidonSponge {
+    // Sponge parameters
+    state: Vec<Zq>,  // Current state of the sponge
+    rate: usize,     // Rate (number of elements to absorb/squeeze)
+    capacity: usize, // Capacity (for cryptographic security)
+    // Poseidon parameters
+    full_rounds: usize, // Number of full rounds (S-box is applied to every element of the state)
+    partial_rounds: usize, // Number of partial rounds (S-box is applied to first element of the state)
+    alpha: u64,            // Exponent for the S-box
+    mds: Vec<Vec<Zq>>,     // MDS matrix
+    ark: Vec<Vec<Zq>>,     // Round constants
+}
+
+impl PoseidonSponge {
+    // Initialize Poseidon sponge parameters
+    pub fn new(
+        rate: usize,
+        capacity: usize,
+        full_rounds: usize,
+        partial_rounds: usize,
+        alpha: u64,
+        mds: Vec<Vec<Zq>>,
+        ark: Vec<Vec<Zq>>,
+    ) -> Self {
+        // State starts with a vector of zeros
+        let state = vec![Zq::ZERO; rate + capacity];
+        Self {
+            state,
+            rate,
+            capacity,
+            full_rounds,
+            partial_rounds,
+            alpha,
+            mds,
+            ark,
+        }
+    }
+
+    // Apply the round constants (ARK) (against algebraic attacks)
+    fn apply_ark(&self, state: &mut [Zq], round_num: usize) {
+        for (i, elem) in state.iter_mut().enumerate() {
+            // adds diffusion (non-linearity)
+            *elem += self.ark[round_num][i];
+        }
+    }
+
+    // Apply the S-box (x -> x^alpha)
+    fn apply_s_box(&self, state: &mut [Zq], is_full_round: bool) {
+        if is_full_round {
+            for elem in state {
+                // Apply transformation to each element of state
+                *elem = elem.pow(self.alpha);
+            }
+        } else {
+            // Apply transformation only to the first element
+            state[0] = state[0].pow(self.alpha);
+        }
+    }
+
+    // Apply the MDS (Maximum Distance Separable) matrix
+    fn apply_mds(&self, state: &mut [Zq]) {
+        let mut new_state = vec![Zq::ZERO; state.len()];
+
+        // Matrix multiplication with the state for diffusion
+        for (i, _cur) in state.iter().enumerate() {
+            let mut sum = Zq::ZERO;
+            for (j, elem) in state.iter().enumerate() {
+                // matrix multiplication
+                sum += *elem * self.mds[i][j];
+            }
+            new_state[i] = sum;
+        }
+
+        state.copy_from_slice(&new_state);
```

Comment:
The ⁠`apply_mds` function creates a new vector and then copies it back, which is inefficient. Use a temporary array on the stack if size is known, or optimize the in-place calculation.

```rs
// After (if size is known at compile time)
fn apply_mds<const N: usize>(&self, state: &mut [Zq; N]) {
    let mut temp = [Zq::ZERO; N];
    // Matrix multiplication
    // ...
    state.copy_from_slice(&temp);
}

// Or with less allocation
fn apply_mds(&self, state: &mut [Zq]) {
    let mut temp = state.to_vec();
    // Matrix multiplication into temp
    // ...
    state.copy_from_slice(&temp);
}
```